File: frontend/src/lib/repairInlineMath.ts
/** 合并被空行/换行拆开的行内公式，避免 Markdown→TipTap 把 $...$ 单独成段。 */

const INLINE_MATH_ONLY = /^\s*\$(?!\$)(.+?)(?<!\$)\$\s*$/;
const HEADING = /^\s*#{1,6}\s/;
const BLOCK_MATH = /^\s*\$\$/;
const ORPHAN_PUNCT = /^[，。；：、」）\)\]】]/;

export function repairFragmentedInlineMath(markdown: string): string {
  if (!markdown || !markdown.includes("$")) return markdown;

  const lines = markdown.replace(/\r\n/g, "\n").split("\n");
  const out: string[] = [];
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];
    const stripped = line.trim();

    if (!stripped) {
      if (
        out.length > 0 &&
        INLINE_MATH_ONLY.test(out[out.length - 1].trim()) &&
        i + 1 < lines.length &&
        lines[i + 1].trim() &&
        !HEADING.test(lines[i + 1]) &&
        !BLOCK_MATH.test(lines[i + 1])
      ) {
        i += 1;
        continue;
      }
      if (
        out.length > 0 &&
        out[out.length - 1].trim() &&
        !INLINE_MATH_ONLY.test(out[out.length - 1].trim()) &&
        i + 1 < lines.length &&
        INLINE_MATH_ONLY.test(lines[i + 1].trim())
      ) {
        i += 1;
        out[out.length - 1] = `${out[out.length - 1].trimEnd()} ${lines[i].trim()}`;
        i += 1;
        continue;
      }
      out.push(line);
      i += 1;
      continue;
    }

    if (BLOCK_MATH.test(stripped) || HEADING.test(stripped)) {
      out.push(line);
      i += 1;
      continue;
    }

    if (INLINE_MATH_ONLY.test(stripped)) {
      if (out.length > 0 && out[out.length - 1].trim() && !INLINE_MATH_ONLY.test(out[out.length - 1].trim())) {
        out[out.length - 1] = `${out[out.length - 1].trimEnd()} ${stripped}`;
      } else if (i + 1 < lines.length && lines[i + 1].trim() && !HEADING.test(lines[i + 1])) {
        const nextStripped = lines[i + 1].trim();
        if (!INLINE_MATH_ONLY.test(nextStripped) && !BLOCK_MATH.test(nextStripped)) {
          out.push(stripped + nextStripped);
          i += 2;
          continue;
        }
        out.push(stripped);
      } else {
        out.push(line);
      }
      i += 1;
      continue;
    }

    if (out.length > 0 && ORPHAN_PUNCT.test(stripped) && out[out.length - 1].trim()) {
      out[out.length - 1] = out[out.length - 1].trimEnd() + stripped;
      i += 1;
      continue;
    }

    out.push(line);
    i += 1;
  }

  return out.join("\n");
}

File: backend/app/services/repair_inline_math.py
"""合并被空行/换行拆开的行内公式，避免 Markdown→TipTap 把 $...$ 单独成段。"""

from __future__ import annotations

import re

_INLINE_MATH_ONLY = re.compile(r"^\s*\$(?!\$)(.+?)(?<!\$)\$\s*$")
_HEADING = re.compile(r"^\s*#{1,6}\s")
_BLOCK_MATH = re.compile(r"^\s*\$\$")
_ORPHAN_PUNCT = re.compile(r"^[，。；：、」）\)\]】]")


def repair_fragmented_inline_math(markdown: str) -> str:
    if not markdown or "$" not in markdown:
        return markdown

    lines = markdown.replace("\r\n", "\n").split("\n")
    out: list[str] = []
    i = 0
    while i < len(lines):
        line = lines[i]
        stripped = line.strip()

        if not stripped:
            if (
                out
                and _INLINE_MATH_ONLY.match(out[-1].strip())
                and i + 1 < len(lines)
                and lines[i + 1].strip()
                and not _HEADING.match(lines[i + 1])
                and not _BLOCK_MATH.match(lines[i + 1])
            ):
                i += 1
                continue
            if (
                out
                and out[-1].strip()
                and not _INLINE_MATH_ONLY.match(out[-1].strip())
                and i + 1 < len(lines)
                and _INLINE_MATH_ONLY.match(lines[i + 1].strip())
            ):
                i += 1
                out[-1] = out[-1].rstrip() + " " + lines[i].strip()
                i += 1
                continue
            out.append(line)
            i += 1
            continue

        if _BLOCK_MATH.match(stripped) or _HEADING.match(stripped):
            out.append(line)
            i += 1
            continue

        if _INLINE_MATH_ONLY.match(stripped):
            if out and out[-1].strip() and not _INLINE_MATH_ONLY.match(out[-1].strip()):
                out[-1] = out[-1].rstrip() + " " + stripped
            elif i + 1 < len(lines) and lines[i + 1].strip() and not _HEADING.match(lines[i + 1]):
                next_stripped = lines[i + 1].strip()
                if not _INLINE_MATH_ONLY.match(next_stripped) and not _BLOCK_MATH.match(next_stripped):
                    out.append(stripped + next_stripped)
                    i += 2
                    continue
                out.append(stripped)
            else:
                out.append(line)
            i += 1
            continue

        if out and _ORPHAN_PUNCT.match(stripped) and out[-1].strip():
            out[-1] = out[-1].rstrip() + stripped
            i += 1
            continue

        out.append(line)
        i += 1

    return "\n".join(out)
